lost/src/main/java/com/mapzen/android/lost/internal/LostClientManager.java

```diff
@@ -269,4 +289,75 @@ public void shutdown() {
+      toSend.putExtra(LocationResult.EXTRA_LOCATION_RESULT, result);
+      intent.send(context, 0, toSend);
+    } catch (PendingIntent.CanceledException e) {
+      Log.e(TAG, "Unable to send pending intent: " + intent);
+    }
+  }
+
+  private void notifyCallback(LostApiClient client, final LocationCallback callback,
+      final LocationResult result) {
+    Looper looper = clientCallbackToLoopers.get(client).get(callback);
+    Handler handler = new Handler(looper);
+    handler.post(new Runnable() {
+      @Override public void run() {
+        callback.onLocationResult(result);
+      }
+    });
+  }
+
+  private void notifyAvailability(LostApiClient client, final LocationCallback callback,
+      final LocationAvailability availability) {
+    Looper looper = clientCallbackToLoopers.get(client).get(callback);
+    Handler handler = new Handler(looper);
+    handler.post(new Runnable() {
+      @Override public void run() {
+        callback.onLocationAvailability(availability);
+      }
+    });
+  }
+
+  private <T> ReportedChanges iterateAndNotify(Location location,
+      Map<LostApiClient, Set<T>> clientToObj, Map<T, LocationRequest> objToLocationRequest,
+      Notifier notifier) {
+    Map<LocationRequest, Long> updatedRequestToReportedTime = new HashMap<>();
+    Map<LocationRequest, Location> updatedRequestToReportedLocation = new HashMap<>();
+    for (LostApiClient client : clientToObj.keySet()) {
+      if (clientToObj.get(client) != null) {
+        for (final T obj : clientToObj.get(client)) {
+          LocationRequest request = objToLocationRequest.get(obj);
+          Long lastReportedTime = requestToLastReportedTime.get(request);
+          Location lastReportedLocation = requestToLastReportedLocation.get(request);
+          if (lastReportedTime == null && lastReportedLocation == null) {
+            updatedRequestToReportedTime.put(request, System.currentTimeMillis());
+            updatedRequestToReportedLocation.put(request, location);
+            notifier.notify(client, obj);
+          } else {
+            long intervalSinceLastReport = System.currentTimeMillis() - lastReportedTime;
+            long fastestInterval = request.getFastestInterval();
+            float smallestDisplacement = request.getSmallestDisplacement();
+            float displacementSinceLastReport = location.distanceTo(lastReportedLocation);
+            if (intervalSinceLastReport >= fastestInterval &&
+                displacementSinceLastReport >= smallestDisplacement) {
+              updatedRequestToReportedTime.put(request, System.currentTimeMillis());
+              updatedRequestToReportedLocation.put(request, location);
+              notifier.notify(client, obj);
+            }
+          }
+        }
+      }
+    }
+    return new ReportedChanges(updatedRequestToReportedTime, updatedRequestToReportedLocation);
+  }
+
+  abstract class Notifier<T> {
+    abstract void notify(LostApiClient client, T obj);
```

Comment:
👍 

I really like the use of generics to allows us to treat `LocationListener`, `PendingIntent`, and `LocationCallback` objects in a similar fashion when sending location updates.